torch/utils/tensorboard/pytorch_graph.py

```diff
@@ -0,0 +1,274 @@
+
+        # torch._C._jit_pass_canonicalize_ops(graph)
+        torch._C._jit_pass_lint(graph)
+
+        torch._C._jit_pass_peephole(graph, True)
+        torch._C._jit_pass_lint(graph)
+
+        # onnx only supports tensors, but 1 / 2 = 0.5 and tensor(1) / tensor(2) = 0
+        torch._C._jit_pass_prepare_division_for_onnx(graph)
+        # onnx only supports tensors, so we turn all out number types into tensors
+        torch._C._jit_pass_erase_number_types(graph)
+        # onnx does not support tuples, so try to remove them
+        torch._C._jit_pass_lower_all_tuples(graph)
+        torch._C._jit_pass_peephole(graph, True)
+        torch._C._jit_pass_lint(graph)
+
+        if operator_export_type != torch.onnx.utils.OperatorExportTypes.RAW:
+            graph = torch._C._jit_pass_onnx(graph, operator_export_type)
+            torch._C._jit_pass_lint(graph)
+            # torch._C._jit_pass_onnx_peephole(graph)
+            torch._C._jit_pass_lint(graph)
+        torch._C._jit_pass_dce(graph)
+        torch._C._jit_pass_lint(graph)
+        torch._C._jit_pass_fixup_onnx_loops(graph)
+        torch._C._jit_pass_lint(graph)
+        graph = torch._C._jit_pass_canonicalize(graph)
+        torch._C._jit_pass_lint(graph)
+        return graph
+
+    assert LooseVersion(torch.__version__) >= LooseVersion("1.0.0"),\
+        'This version of tensorboardX requires pytorch>=1.0.0.'
+
+    with torch.onnx.set_training(model, False):
+        try:
+            trace, _ = torch.jit.get_trace_graph(model, args)
+        except RuntimeError:
+            print('Error occurs, No graph saved')
+            _ = model(*args)  # don't catch, just print the error message
+            print("Checking if it's onnx problem...")
+            try:
+                import tempfile
+                torch.onnx.export(
+                    model, args, tempfile.TemporaryFile(), verbose=True)
+            except RuntimeError:
+                print("Your model fails onnx too, please report to onnx team")
+            return GraphDef(versions=VersionDef(producer=22))
+
+    if 'operator_export_type' not in kwargs:
+        operator_export_type = torch.onnx.utils.OperatorExportTypes.ONNX
+    else:
+        operator_export_type = getattr(torch.onnx.utils.OperatorExportTypes, kwargs['operator_export_type'])
+
+    if 'omit_useless_nodes' not in kwargs:
+        omit_useless_nodes = True
+
+    try:
+        _optimize_trace(trace, operator_export_type)
+    except RuntimeError as e:
+        logging.warn(ImportError(e))
+    graph = trace.graph()
+    if verbose:
+        print(graph)
+    list_of_nodes, node_stats = parse(graph, args, omit_useless_nodes)
+    stepstats = RunMetadata(step_stats=StepStats(dev_stats=[DeviceStepStats(device="/device:CPU:0",
+                                                                            node_stats=node_stats)]))
+    return GraphDef(node=list_of_nodes, versions=VersionDef(producer=22)), stepstats
```

Comment:
How did we compute the number 22? Under what circumstances can we update it?

Reply:
I guess, it would be helpful to link to the canonical documentation for GraphDef. Is https://github.com/tensorflow/tensorflow/blob/r1.13/tensorflow/core/framework/graph.proto appropriate?

Reply:
Should be: https://github.com/tensorflow/tensorflow/blob/d6d5751a3c046dd38a9998a2db495c7e8f426ff6/tensorflow/core/public/version.h#L108

Reply:
Actually going to link https://github.com/tensorflow/tensorboard/blob/master/tensorboard/compat/proto/graph.proto if people are okay with it.